### ISSUE TITLE
[RFC] HACPO: heterogeneous-agent collaborative training on verl V1

### ISSUE THREAD

**Issue by JacckMa** [verl-project/verl-recipe#143]
# [RFC] HACPO: heterogeneous-agent collaborative training on verl V1

## Summary

This RFC proposes a [HACPO](https://arxiv.org/pdf/2603.02604) recipe for jointly training two heterogeneous language models on verl V1. For each shared prompt batch, the policies independently generate responses, and HACPO uses the resulting joint set of verified rollouts to update every policy while accounting for capability and policy-distribution discrepancies.

The implementation follows the [official HACPO code](https://github.com/zzx-peter/HACRL-code), while keeping all HACPO-specific runtime and loss logic inside `verl-recipe`. No verl core changes are required.

## Motivation

When heterogeneous policies are trained independently with RLVR, each verified rollout updates only the policy that generated it. HACPO lets two heterogeneous policies reuse each other's rollouts during training while remaining independent at inference time, improving sample utilization and enabling bidirectional knowledge transfer.

The official implementation is built on an older version of verl and uses a fixed main/aux model layout. Porting HACPO to V1 requires more than a loss or configuration change: both policies need independent rollout and optimizer state, and a response generated with one tokenizer must be retokenized before another policy can evaluate its likelihood.

<img width="1936" height="1206" alt="Image" src="https://github.com/user-attachments/assets/5b3db17c-48ec-4c28-92e3-a14079a55627" />

*HACPO uses the joint set of verified rollouts from heterogeneous policies to update every policy, with capability- and distribution-aware corrections for stable collaborative training. Figure from the HACRL paper.*

## Proposal

The recipe adds three capabilities through existing verl V1 extension points:

- **Multi-policy orchestration:** a recipe-local V1 `PPOTrainer` subclass for synchronous training maintains a policy-indexed set of worker groups, rollout servers, agent loops, and checkpoint managers. Each policy generates responses for the shared prompt batch, and every policy is then updated from the joint verified rollout set. The runtimes share one resource pool and use verl's existing sleep and offload mechanisms between phases.
- **Cross-policy training batches:** the controller retains canonical response text and each source policy's length-normalized response log-probability in a tokenizer-neutral trajectory pool. Cross-policy responses are retokenized with the learner's tokenizer before being materialized as V1 `TensorDict` batches.
- **HACPO objective integration:** the HACPO loss is registered through the V1 model-engine loss interface. A recipe-local `TrainingWorker` subclass supplies the optimizer mini-batch index required by stepwise clipping. Together with controller-side advantage estimation, this implements the four HACPO mechanisms: Agent-Capability-Aware Advantage Estimation, the Model Capabilities Discrepancy Coefficient, Exponential Importance Sampling, and Stepwise Clipping.

The recipe also provides a Math-Verify reward, configuration, launch presets, and focused tests. All HACPO-specific implementation is contained under `hacpo/`.

### Training flow

For each training step:

1. The controller draws one prompt batch and activates each policy in turn to generate and verify its rollouts.
2. The resulting trajectories are merged into a tokenizer-neutral pool that preserves prompt identity, source policy, reward, response text, and source sequence log-probability.
3. For each learner, HACPO computes capability-aware advantages from the joint reward set. Self-generated trajectories retain their native tokenization, while cross-policy responses are retokenized with the learner's tokenizer.
4. The policies update in turn with the HACPO loss while inactive runtimes sleep or offload their state. Metrics remain policy-specific, and checkpoints store both policies under the same global step.

## Data and evaluation

The recipe trains on the paper's 7.5k MATH set and validates on MATH-500, with launch presets for Qwen3-4B-Base + Qwen3-1.7B-Base and Qwen3-4B-Base + Llama-3.2-3B-Instruct. Training and the curves below use a maximum response length of **4096**.

The recipe follows the two-agent setup evaluated in the paper, while its configuration and orchestration interfaces leave room for future extensions to more agents.

## Training curves

The following curves report greedy MATH-500 `avg@1` during HACPO training of Qwen3-4B-Base and Qwen3-1.7B-Base. Both use a maximum response length of **4096**, validation every three steps, the paper's prompt template with thinking disabled, and seed 42.

<img width="2384" height="714" alt="Image" src="https://github.com/user-attachments/assets/66e76e28-2b34-4ae3-8146-3ff1771d0081" />

<img width="2394" height="730" alt="Image" src="https://github.com/user-attachments/assets/35c9080c-c49f-4ab2-bed8-2b8793967b57" />

The cross-tokenizer Qwen3-4B-Base and Llama-3.2-3B-Instruct run uses the same validation protocol.

<img width="2376" height="710" alt="Image" src="https://github.com/user-attachments/assets/b48aa162-daa8-41b4-b229-dbd43fd12b27" />

<img width="2386" height="712" alt="Image" src="https://github.com/user-attachments/assets/d8504325-1cf9-49b0-b568-ea6027955989" />

## References

- Paper: https://arxiv.org/abs/2603.02604
- Paper page: https://huggingface.co/papers/2603.02604
- Official implementation: https://github.com/zzx-peter/HACRL-code
